### DISCUSSION TITLE
Duplicated timeseries in CollectorRegistry

### DISCUSSION THREAD

**Discussion by jednoprsak** [prometheus/client_python#924] [Q&A]
Hello,
I am trying to make my metrics and write it inside one file using textfile_collector, but the problem is, that it tells me ''Duplicated timeseries in CollectorRegistry: {'htcondor_jobstatus'}'.

Here is the function:
```
def make_job_metrics():
    # works with mandatory structures 
    # JobsAd_metrics, JobsAd_labels, jobs, metrics
    registry = CollectorRegistry()
    for jobid in jobs:
      # print(type(jobid))
      if not isinstance(jobid, str):
          continue
      for metric in JobsAd_metrics:
        args_dict = {}
        for parameter in JobsAd_labels:
            args_dict[parameter] = jobs[jobid][parameter]
        metric_name = 'htcondor' + '_' + metric.lower()
        # print(metric_name)
        # print(args_dict)
        print(metric_name)
        args_list=[]
        for v in args_dict.items():
            args_list.append(v[1])
        if jobs[jobid][metric] == None:
            continue
        **g = Gauge(metric_name, 'metric_label', JobsAd_labels, registry=registry).labels(*args_list).set(jobs[jobid][metric])**
    write_to_textfile('/usr/local/bin/prometheus_custom_plugins/test.prom', registry)

```
Here is the command line output:
Traceback (most recent call last):
  File "./htcondor_exporter.py", line 308, in <module>
    make_job_metrics()
  File "./htcondor_exporter.py", line 147, in make_job_metrics
    g = Gauge(metric_name, 'metric_label', JobsAd_labels, registry=registry).labels(*args_list).set(jobs[jobid][metric])
  File "/usr/local/bin/prometheus_custom_plugins/htcenv/lib/python3.6/site-packages/prometheus_client/metrics.py", line 373, in __init__
    _labelvalues=_labelvalues,
  File "/usr/local/bin/prometheus_custom_plugins/htcenv/lib/python3.6/site-packages/prometheus_client/metrics.py", line 143, in __init__
    registry.register(self)
  File "/usr/local/bin/prometheus_custom_plugins/htcenv/lib/python3.6/site-packages/prometheus_client/registry.py", line 45, in register
    duplicates))
ValueError: Duplicated timeseries in CollectorRegistry: {'htcondor_jobstatus'}

My question is: How to write many metrics with same names, but different labels and values inside one file using textfile collector? 
- because it looks like, that in the case when I put inside registry next metric with the same name, it prints out this error.

Thank you for any responses
Lukáš Míča

**Comment by csmarchbanks** [comment]
Hello, typically you will want to create a single `Gauge` for all of the metrics with the same name since they are considered one "Metric Family" by Prometheus. I don't know what your datastructures look like, but my guess is you will want to add a dictionary that maps a metric name to a Gauge and uses that with `.labels(*args_list).set(jobs[jobid][metric])` in your loop. In the case there is no entry in the dictionary then create a new Gauge.

**Reply by jednoprsak** [reply]
Please could you write me simple example, how to create many metrics with the same name, but different values and labels, all without my function?

**Reply by ajaxaddicted** [reply]
For someone else that might struggle with the same issue. This is what a solution might look like:

```python
from prometheus_client import Gauge, Counter, CollectorRegistry
from prometheus_client import exposition

metrics_list = [
    {
        "name": "api_http_requests_total",
        "label_keys": ["method", "handler"],
        "label_values": [
            "POST",
            "/messages",
        ],
        "type": "gauge",
        "value": 10,
    },
    {
        "name": "api_http_requests_total",
        "label_keys": ["method", "handler"],
        "label_values": [
            "GET",
            "/messages/:id",
        ],
        "type": "gauge",
        "value": 15,
    }
]


def generate_prometheus_response(metrics: List):
    registry = CollectorRegistry()

    uniqueness_dict = {}
    for metric in metrics:
        label_keys = metric.get("label_keys", [])
        label_values = metric.get("label_values", [])
        metric_type = metric.get("type", "gauge")
        value = metric.get("value", 0)
        name = metric.get("name", "")
        if name not in uniqueness_dict:
            if metric_type == "counter":
                _generated = Counter(
                    name=name, documentation="", labelnames=label_keys, registry=registry
                )
                _generated.labels(*label_values).inc(value)
                uniqueness_dict[name] = _generated

            if metric_type == "gauge":
                _gauge = Gauge(
                    name=name,
                    documentation="",
                    labelnames=label_keys,
                    registry=registry,
                )
                _gauge.labels(*label_values).set(value)
                uniqueness_dict[name] = _gauge
        else:
            if metric_type == "counter":
                uniqueness_dict[name].labels(*label_values).inc(value)

            if metric_type == "gauge":
                uniqueness_dict[name].labels(*label_values).set(value)

    return exposition.generate_latest(registry=registry)
```

**Reply by jednoprsak** [reply]
Thanks
